runtime/android/core/src/org/xwalk/core/XWalkDialogManager.java

```diff
@@ -164,24 +361,61 @@ private AlertDialog buildAlertDialog() {
         return dialog;
     }
 
-    private void setPositiveButton(AlertDialog dialog, String text, final Runnable command) {
-        dialog.setButton(DialogInterface.BUTTON_POSITIVE, text,
-                new OnClickListener() {
-                    @Override
-                    public void onClick(DialogInterface dialog, int id) {
-                        command.run();
-                    }
-                });
+    private ProgressDialog buildProgressDialog() {
+        ProgressDialog dialog = new ProgressDialog(mContext);
+        dialog.setIndeterminate(true);
+        dialog.setCancelable(false);
+        dialog.setCanceledOnTouchOutside(false);
+        return dialog;
+    }
+
+    private void setTitle(AlertDialog dialog, int resourceId) {
+        dialog.setTitle(mContext.getString(resourceId));
+    }
+
+    private void setMessage(AlertDialog dialog, int resourceId) {
+        dialog.setMessage(replaceApplicationName(mContext.getString(resourceId)));
     }
 
-    private void setNegativeButton(AlertDialog dialog, String text, final Runnable command) {
-        dialog.setButton(DialogInterface.BUTTON_NEGATIVE, text,
-                new OnClickListener() {
-                    @Override
-                    public void onClick(DialogInterface dialog, int id) {
-                        command.run();
+    private void setPositiveButton(AlertDialog dialog, int resourceId) {
+        dialog.setButton(DialogInterface.BUTTON_POSITIVE, mContext.getString(resourceId),
+                (OnClickListener) null);
+    }
+
+    private void setNegativeButton(AlertDialog dialog, int resourceId) {
+        dialog.setButton(DialogInterface.BUTTON_NEGATIVE, mContext.getString(resourceId),
+                (OnClickListener) null);
+    }
+
+    private void showDialog(final AlertDialog dialog, final ArrayList<ButtonAction> actions) {
+        dialog.setOnShowListener(new OnShowListener() {
+            @Override
+            public void onShow(DialogInterface d) {
+                for (ButtonAction action : actions) {
+                    Button button = dialog.getButton(action.mWhich);
+                    if (button == null) {
+                        if (action.mMandatory) {
+                            throw new RuntimeException("Button " + action.mWhich + " is mandatory");
```

Comment:
`mWhich`'s value is `BUTTON_POSITIVE` or `BUTTON_NEGATIVE`, they are integer, is it make sense to user?